### PR TITLE
Feat: new powershell connector

### PR DESCRIPTION
**Pull request by jbreek**
Migrated to new powershell connector

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,29 +160,25 @@ Besides the configuration tab, you can also configure script variables. To decid
...
   $titleLookupProperty = { $_.Title.ExternalId }
 ```
 
-
-
-
 ## Getting help
 
-> _For more information on how to configure a HelloID PowerShell connector, please refer to our [documentation](https://docs.helloid.com/hc/en-us/articles/360012558020-Configure-a-custom-PowerShell-target-system) pages_
+> [!TIP] > _For more information on how to configure a HelloID PowerShell connector, please refer to our [documentation](https://docs.helloid.com/en/provisioning/target-systems/powershell-v2-target-systems.html) pages_.
```
Comment (rhouthuijzen) [inline README.md:179]: Markdown is broken. I guess the 2nd ">" needs to be on a new line

Comment (rhouthuijzen) [inline README.md]: Markdown is broken. I guess the 2nd ">" needs to be on a new line

File: README.md [hunk 2]
```diff
@@ -137,29 +160,25 @@ Besides the configuration tab, you can also configure script variables. To decid
...
 
-> _If you need help, feel free to ask questions on our [forum](https://forum.helloid.com/forum/helloid-connectors/provisioning/4792-helloid-conn-prov-target-inception)
+> [!TIP] > \_If you need help, feel free to ask questions on our [forum](https://forum.helloid.com/forum/helloid-connectors/provisioning/4792-helloid-conn-prov-target-inception)
```
Comment (rhouthuijzen) [inline README.md]: Markdown is broken. I guess the 2nd ">" needs to be on a new line

File: README.md [hunk 3]
```diff
@@ -1,51 +1,78 @@
...
+
+2. Specify the following configuration:
+
+   | Setting                   | Value                             |
+   | ------------------------- | --------------------------------- |
+   | Enable correlation        | `True`                            |
+   | Person correlation field  | `PersonContext.Person.ExternalId` |
+   | Account correlation field | `staffnumber`                     |
+
+> [!TIP] > _For more information on correlation, please refer to our correlation [documentation](https://docs.helloid.com/en/provisioning/target-systems/powershell-v2-target-systems/correlation.html) pages_.
```
Comment (rhouthuijzen) [inline README.md]: Markdown is broken. I guess the 2nd ">" needs to be on a new line

File: config.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
+[
```
Comment (rhouthuijzen) [inline config.json]: Is this file still needed?

File: create.ps1 [hunk 5]
```diff
@@ -378,60 +235,154 @@ function Get-InceptionIdsFromHelloIdContact {
...
-            throw "More than one active account was found for [$($account.staffnumber)]. Please remove the obsolete account(s) or make sure that the obsolete account(s) are disabled and enable only the correct account."
-        }
-    }
+    # Properties Excluded from response
+    $excludedProperties = "dateofbirth,phone,mobile,residence,state,supervisorid,rolesPerOrgUnits,positionsPerOrgGroup,rolesPerOrgGroup"
+    $excludedPropertiesArray = $excludedProperties -split ","
 
-    $aRef = ($resultGetEmployee.items | Select-Object -First 1).id
+    # Validate correlation configuration
+    if ($actionContext.CorrelationConfiguration.Enabled) {
```
Comment (rhouthuijzen) [inline create.ps1:327]: Is correlation mandatory? If it is required please add a throw.

For example:
```Powershell
    else {
        $outputContext.AuditLogs.Add([PSCustomObject]@{
                Action  = "CorrelateAccount"
                Message = "Configuration of correlation is madatory."
                IsError = $true
            })
        Throw "Configuration of correlation is madatory."
    }
```

File: README.md [hunk 6]
```diff
@@ -1,51 +1,78 @@
...
+To properly setup the correlation:
+
+1. Open the `Correlation` tab.
+
+2. Specify the following configuration:
+
+   | Setting                   | Value                             |
+   | ------------------------- | --------------------------------- |
+   | Enable correlation        | `True`                            |
+   | Person correlation field  | `PersonContext.Person.ExternalId` |
```
Comment (rhouthuijzen) [inline README.md:68]: In your script you use the accountFieldValue for correlation. The personFieldValue is not used. So this should be empty.

File: create.ps1 [hunk 7]
```diff
@@ -378,60 +235,154 @@ function Get-InceptionIdsFromHelloIdContact {
...
+        if ([string]::IsNullOrEmpty($($correlationField))) {
+            throw 'Correlation is enabled but not configured correctly'
+        }
+        if ([string]::IsNullOrEmpty($($correlationValue))) {
+            throw 'Correlation is enabled but [accountFieldValue] is empty. Please make sure it is correctly mapped'
+        }
+
+        # Verify if a user must be either [created ] or just [correlated]
+        $headers = @{
+            Accept        = 'application/json'
```
Comment (rhouthuijzen) [inline create.ps1:340]: Is this tested with special characters, for example, é?

In some cases you need to use: 'application/json; charset=utf-8'

File: create.ps1 [hunk 8]
```diff
@@ -445,149 +396,111 @@ try {
...
     # Process
-    if (-not($dryRun -eq $true)) {
+    if (-not($actionContext.DryRun -eq $true)) {
         switch ($action) {
-            'Create-Correlate' {
-                Write-Verbose 'Creating and correlating Inception Employee account'
+            'CreateAccount' {                
+                Write-Verbose 'Gathering Inception Positions and organization Units to map the against the HelloId person'
+                $positions = Get-InceptionPosition -Headers $headers -pageSize 1000
+                $orgUnits = Get-InceptionOrgunit -Headers $headers -pageSize 1000
```
Comment (rhouthuijzen) [inline create.ps1]: The code in rows 414-416 is the same as in rows 388-390. Is this the intention?

File: create.ps1 [hunk 9]
```diff
@@ -378,60 +235,154 @@ function Get-InceptionIdsFromHelloIdContact {
...
+        $action = 'CreateAccount'
+    }
+    else {
+        $action = 'CorrelateAccount'
+        $outputContext.AccountReference = $correlatedAccount.id
+    }
+
+    # Add a message and the result of each of the validations showing what will happen during enforcement
+    if ($actionContext.DryRun -eq $true) {
+        Write-Verbose "[DryRun] Inception $action for: [$($personContext.Person.DisplayName)], will be executed during enforcement" -Verbose
```
Comment (rhouthuijzen) [inline create.ps1]: Debug toggle implemented please search your code for Write-Verbose -Verbose and remove it. Or replace write-verbose by write-information or write-warning.

File: create.ps1 [hunk 10]
```diff
@@ -1,189 +1,22 @@
...
-    email                = $p.Contact.Business.Email
-    phone                = $p.Contact.Business.Phone.Fixed
-    dateofbirth          = if ($null -ne $p.Details.BirthDate ) { '{0:yyyy-MM-dd}' -f ([datetime]$p.Details.BirthDate) };
-    startdate            = if ($null -ne $p.PrimaryContract.StartDate ) { '{0:yyyy-MM-dd}' -f ([datetime]$p.PrimaryContract.StartDate) };
-    enddate              = $null # ((Get-Date).AddDays(-1))
-    positionsPerOrgUnits = @()
-}
-
+# PowerShell V2
+# Version: 1.0.0
```
Comment (rhouthuijzen) [inline create.ps1]: Version needs to be 2.0.0. But can also be removed (like in template)

The PSv1 will be "release" v1.0.2. This code should be "release" v2.0.0

File: fieldMapping.json [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+      "Description": "",
+      "Type": "Text",
+      "MappingActions": [
+        {
+          "MapForActions": [
+            "Create",
+            "Update"
+          ],
+          "MappingMode": "Complex",
+          "Value": "\"function GetEmailAddress(){\\r\\n    return Person.Accounts._bccd667807a9431a9d4acfc64c19a3d6.mail\\r\\n}\\r\\nGetEmailAddress()\"",
```
Comment (rhouthuijzen) [inline fieldMapping.json]: Please change the GUID by MicrosoftActiveDirectory.

I recommend checking if the value is avaliable:

```JavaScript

function getEmail() {
    let mail = '';

    if (typeof Person.Accounts.MicrosoftActiveDirectory.mail !== 'undefined' && Person.Accounts.MicrosoftActiveDirectory.mail) {
        mail = Person.Accounts.MicrosoftActiveDirectory.mail;
    }

    return mail;
}

getEmail()
```

File: fieldMapping.json [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+      "Description": "",
+      "Type": "Text",
+      "MappingActions": [
+        {
+          "MapForActions": [
+            "Create",
+            "Update"
+          ],
+          "MappingMode": "Complex",
+          "Value": "\"const formatedBirthDatestamp = () => {\\r\\n    const birthDate = Person.Details.BirthDate\\r\\n    const date = birthDate.toISOString().split('T')[0];\\r\\n    const time = \\\"00:00:00\\\" //birthDate.toTimeString().split(' ')[0].replace(/:/g, ':');\\r\\n    return `${date}T${time}Z`\\r\\n}\\r\\n\\r\\nfunction GetFormattedBirthDate(){\\r\\n    let formattedDate = formatedBirthDatestamp();\\r\\n    return formattedDate\\r\\n}\\r\\nGetFormattedBirthDate()\"",
```
Comment (rhouthuijzen) [inline fieldMapping.json]: Did you check what happens if the BirthDate is empty? Maybe a the following code needs to be added to make sure the javescript won't fail.

```JavaScript
if (typeof Person.Details.BirthDate !== 'undefined' && Person.Details.BirthDate) {

}
```

File: resources.ps1 [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+            $auditMessage = "Could not create Inception resource. Error: $($ex.Exception.Message)"
+            Write-Verbose "Error at Line '$($ex.InvocationInfo.ScriptLineNumber)': $($ex.InvocationInfo.Line). Error: $($ex.Exception.Message)"
+        }
+        $outputContext.AuditLogs.Add([PSCustomObject]@{
+                Message = $auditMessage
+                IsError = $true
+            })
+    }    
+}
+catch {
```
Comment (rhouthuijzen) [inline resources.ps1:315]: Can you explain the purpose of the nested try/catch?

File: update.ps1 [hunk 14]
```diff
@@ -376,146 +190,287 @@ function Get-InceptionIdsFromHelloIdContact {
...
+            $httpErrorObj.FriendlyMessage = $httpErrorObj.ErrorDetails # Temporarily assignment
+        }
+        catch {
+            $httpErrorObj.FriendlyMessage = $httpErrorObj.ErrorDetails
+        }
+        Write-Output $httpErrorObj
+    }
+}
+#endregion
+$actionContext.DryRun = $false
```
Comment (rhouthuijzen) [inline update.ps1]: Leftover code from testing?

File: create.ps1 [hunk 15]
```diff
@@ -445,149 +396,111 @@ try {
...
-            IsError = $false
-        })
 
+        $accountReferenceObject = @{
+            AccountReference = $accountReferenceId
+            Positions        = $desiredPositionList | Select-Object * -ExcludeProperty SideIndicator
+        }
+
+        $outputContext.AccountReference = $accountReferenceObject
+        $outputContext.success = $true
```
Comment (rhouthuijzen) [inline create.ps1]: DryRun will never make success = $true.

I recommend adding a finaly to your scripts (by default  $outputContext.Success = $false). For example:

```Powershell
finally {
    # Check if auditLogs contains errors, if no errors are found, set success to true
    if (-not($outputContext.AuditLogs.IsError -contains $true)) {
        $outputContext.Success = $true
    }
}
```

**Review by rhouthuijzen — COMMENTED** [review]
Overall code looks good. I cannot estimate whether the code works. I mainly made some suggestions and found some small errors.

If you have any questions please let me know.

File: create.ps1 (hunk 8 above)
Comment (rhouthuijzen) [reply create.ps1]: Additional question. Is there a reason why the pageSize is different from the default '200'?

File: resources.ps1 [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+            Accept        = 'application/json'
+            Authorization = "Bearer $(Get-InceptionToken)"
+        }
+
+        $OrgUnits = Get-InceptionOrgunits -Headers $headers
+        $positions = Get-InceptionPositions -Headers $headers
+
+
+        foreach ($resource in $resourceContext.SourceData) {
+            if ($True) {
```
Comment (rhouthuijzen) [inline resources.ps1]: This if statement is unnecessary?

File: enable.ps1 [hunk 17]
```diff
@@ -1,48 +1,41 @@
...
     $false { $VerbosePreference = 'SilentlyContinue' }
 }
 
-function Get-RandomCharacters($length, $characters) { 
-    $random = 1..$length | ForEach-Object { Get-Random -Maximum $characters.length }
-    return [String]$characters[$random]
-}
+# Script Variables
+#$employeeFound = $false
+#$userFound     = $false
```
Comment (rhouthuijzen) [inline enable.ps1]: $employeeFound and $userFound can be removed? (use ctrl F to check code)

File: enable.ps1 [hunk 18]
```diff
@@ -64,183 +57,175 @@ function Resolve-InceptionError {
...
-# Begin
 try {
-    Write-Verbose "Verifying if a Inception account for [$($p.DisplayName)] exists"
+    # Verify if [aRef] has a value
+    if ([string]::IsNullOrEmpty($($actionContext.References.Account))) {
+        throw 'The account reference could not be found'
+    }
+
+    Write-Verbose "Verifying if an Inception account for [$($personContext.Person.DisplayName)] (still) exists"
+    #$correlatedAccount = 'userInfo'
```
Comment (rhouthuijzen) [inline enable.ps1]: $correlatedAccount = 'userInfo' can be removed in the scripts (left over from template)

File: update.ps1 [hunk 19]
```diff
@@ -1,233 +1,48 @@
...
 # Set debug logging
-switch ($($config.IsDebug)) {
+switch ($($actionContext.Configuration.isDebug)) {
     $true { $VerbosePreference = 'Continue' }
     $false { $VerbosePreference = 'SilentlyContinue' }
 }
 
+# Script Configuration
+$departmentLookupProperty = { $_.Department.ExternalId }
+$titleLookupProperty = { $_.Title.ExternalId }
```
Comment (rhouthuijzen) [inline update.ps1:21]: This part of the code is complex. Not for you to change. This should be something for the connector team.

File: README.md (hunk 1 above)
Comment (jbreek) [reply README.md]: Corrected the markdown

File: README.md (hunk 2 above)
Comment (jbreek) [reply README.md]: Corrected the markdown

File: README.md (hunk 3 above)
Comment (jbreek) [reply README.md]: Corrected the markdown

File: config.json (hunk 4 above)
Comment (jbreek) [reply config.json]: Removed obsolete file

File: create.ps1 (hunk 5 above)
Comment (jbreek) [reply create.ps1:327]: Added throw

File: README.md (hunk 6 above)
Comment (jbreek) [reply README.md:68]: These are the settings on the correlation tab of the targetconnector.

File: create.ps1 (hunk 7 above)
Comment (jbreek) [reply create.ps1:340]: Yes it is tested. The charset is added to the contenttype of the POST action to create the employee or the user.

File: create.ps1 (hunk 8 above)
Comment (jbreek) [reply create.ps1]: Duplicate code removed, was used for testing. PageSize set to default. Could be for performance to get greater pagesize.

File: create.ps1 (hunk 9 above)
Comment (jbreek) [reply create.ps1]: Removed

File: create.ps1 (hunk 10 above)
Comment (jbreek) [reply create.ps1]: updated version

File: fieldMapping.json (hunk 12 above)
Comment (jbreek) [reply fieldMapping.json]: Added check on existance.

File: resources.ps1 (hunk 13 above)
Comment (jbreek) [reply resources.ps1:315]: Think this the result of copy/paste from original script to template. Removed the nested try/catch

File: update.ps1 (hunk 14 above)
Comment (jbreek) [reply update.ps1]: Removed

File: create.ps1 (hunk 15 above)
Comment (jbreek) [reply create.ps1]: Implemented in scripts

File: resources.ps1 (hunk 16 above)
Comment (jbreek) [reply resources.ps1]: Don't know why this is in there. Removed it cause it doesn't seem to have any functionality.

File: enable.ps1 (hunk 17 above)
Comment (jbreek) [reply enable.ps1]: Removed

File: enable.ps1 (hunk 18 above)
Comment (jbreek) [reply enable.ps1]: Removed

File: update.ps1 (hunk 19 above)
Comment (jbreek) [reply update.ps1:21]: This code was implemented by connector team.

**Review by rhouthuijzen — APPROVED** [review]
Code is implemented based on the template. Recommended fixes are resolved.
